File: LKO-ADC/koopman_project/src/lko-lstm/generate_lstm_data.m
function [file_control, file_state, file_labels] = generate_lstm_data(control, states, time_step, pred_step)
    if nargin < 4
        pred_step = 1; % 
    end

    % 提取数据并验证维度
    [c, t] = size(control);
    [d, t_check] = size(states);


    % 数据一致性检查
    if t ~= t_check
        fprintf('时间步不匹配');
        return;
    end
    if t < time_step+1
        fprintf('时间步不足');
        return;
    end

    % 计算本文件样本数
    num_samples = t - time_step - pred_step + 1;  % 保证标签窗口有足够数据

    % 预分配本文件数据
    file_control = zeros(c, pred_step,  num_samples);
    file_state = zeros(d, num_samples, time_step);
    file_labels = zeros(d, time_step, pred_step, num_samples);
    
    % 构建时间窗口
    for sample_idx = 1:num_samples
        time_window = sample_idx : sample_idx + time_step - 1;
        
        % 当前状态序列 [s(t) ... s(t+m-1)]
        file_state(:, sample_idx, :) = states(:, time_window);
        
        for k = 1:pred_step
            % 标签序列 [s(t+1) ... s(t+m)]
            file_labels(:, :, k, sample_idx) = states(:, time_window + k);
            file_control(:, k, sample_idx) = control(:, sample_idx + time_step+k-2);
        end
    end
end

File: LKO-ADC/python_koopman_project/src/DRKO/generate_lstm_data.py
import numpy as np
import torch


def generate_lstm_data(control, states, time_step, pred_step=1):
    # 确保输入是numpy数组
    control = np.asarray(control)
    states = np.asarray(states)

    # 提取维度并进行验证
    c, t = control.shape
    d, t_check = states.shape

    if t != t_check:
        raise ValueError(f"时间维度不匹配: control有 {t} 个时间步, states有 {t_check} 个时间步。")

    # 计算可以生成的样本窗口数量
    # 需要足够的数据来容纳一个历史窗口(time_step)和一个预测窗口(pred_step)
    num_samples = t - time_step - pred_step + 1

    if num_samples < 1:
        raise ValueError(f"时间序列太短，无法满足指定的 time_step ({time_step}) 和 pred_step ({pred_step})。"
                         f"至少需要 {time_step + pred_step} 个数据点。")

    # 预分配NumPy数组以存储结果
    # 形状格式与MATLAB脚本保持一致:
    # file_control: (特征, 预测步数, 样本数, 历史步数)
    # file_state:   (特征, 样本数, 历史步数)
    # file_label:   (特征, 预测步数, 样本数, 历史步数) -> 注意：这是一个不寻常的形状
    file_control = np.zeros((c, pred_step, num_samples, time_step))
    file_state = np.zeros((d, num_samples, time_step))
    file_label = np.zeros((d, pred_step, num_samples, time_step))

    # 通过在时间序列上滑动窗口来生成每个样本
    for i in range(num_samples):
        # 定义输入状态历史的索引窗口。
        history_indices = np.arange(i, i + time_step, 1)
        file_state[:, i, :] = states[:, history_indices]

        for k_idx in range(pred_step):
            k = k_idx + 1  # 使用1-based的k来计算标签的偏移量

            # 标签是状态窗口在未来偏移k步的结果
            label_indices = history_indices + k
            file_label[:, k_idx, i, :] = states[:, label_indices]

            # 控制输入对应于历史窗口的末端
            file_control[:, k_idx, i, :] = control[:, history_indices]

    file_state = np.transpose(file_state, (1, 2, 0))
    file_control = np.transpose(file_control, (2, 1, 3, 0))
    file_label = np.transpose(file_label, (2, 1, 3, 0))

    return file_control, file_state, file_label
